models/concept.go

```diff
@@ -0,0 +1,82 @@
+package models
+
+import (
+	"log"
+
+	"github.com/uc-cdis/cohort-middleware/utils"
+	"gorm.io/gorm"
+)
+
+type Concept struct {
+	ConceptId   int
+	ConceptName string
+	DomainId    int
+	DomainName  string
+}
+
+type ConceptStats struct {
+	ConceptId     int
+	ConceptName   string
+	NmissingRatio float32
+}
+
+type Observation struct {
+}
+
+func getDataSource(sourceId int, schemaName string) *gorm.DB {
+	var dataSourceModel = new(Source)
+	dataSource, _ := dataSourceModel.GetSourceByIdWithConnection(sourceId)
+
+	sourceConnectionString := dataSource.SourceConnection
+	dbSchema := schemaName + "."
+	omopDataSource := utils.GetDataSourceDB(sourceConnectionString, dbSchema)
+	return omopDataSource
+}
+
+func (h Concept) RetriveAllBySourceId(sourceId int) ([]*Concept, error) {
+	omopDataSource := getDataSource(sourceId, "OMOP")
+
+	var concept []*Concept
+	omopDataSource.Model(&Concept{}).
+		Select("concept_id, concept_name, domain_id, '' as domain_name").
+		Order("concept_name").
+		Scan(&concept)
+	return concept, nil
+}
+
+func (h Concept) RetriveStatsBySourceIdAndCohortIdAndConceptIds(sourceId int, cohortDefinitionId int, conceptIds []int) ([]*ConceptStats, error) {
+	omopDataSource := getDataSource(sourceId, "OMOP")
+
+	var conceptStats []*ConceptStats
+	omopDataSource.Model(&Concept{}).
+		Select("concept_id, concept_name, domain_id, '' as domain_name, 0 as n_missing_ratio").
+		Where("concept_id in (?)", conceptIds).
+		Order("concept_name").
+		Scan(&conceptStats)
+
+	resultsDataSource := getDataSource(sourceId, "RESULTS")
+	var cohortSubjectIds []int
+	resultsDataSource.Model(&Cohort{}).
+		Select("subject_id").
+		Where("cohort_definition_id = ?", cohortDefinitionId).
+		Scan(&cohortSubjectIds)
+	var cohortSize = len(cohortSubjectIds)
+
+	// find, for each concept, the ratio of persons in the given cohortId that have
+	// a non-empty value for this concept:
+	for _, conceptStat := range conceptStats {
+
+		var nrPersonsWithData int
+		omopDataSource.Model(&Observation{}).
+			Select("count(distinct(person_id))").
+			Where("observation_concept_id = ?", conceptStat.ConceptId).
+			Where("person_id in (?)", cohortSubjectIds).
+			Where("value_as_string is not null"). // transform to is null
```

Comment:
@pieterlukasse This needs to do `value_as_number is not null OR value_as_string is not null`

Reply:
@kmhernan thinking a bit more about this, should I also count as missing the persons that do not have this concept at all?

Reply:
yes, that's what I'm doing. It just feels wrong because some concepts are not really missing because data is incomplete, but just because the `person` has a different concept in the same domain. E.g. `8507,'MALE'` is a concept, at least in my test db now.... I think my concept of what is a concept might be wrong... 🤔

Reply:
I think I got my concept examples from a vocabulary that is a hierarchy of concepts...I see that the vocabulary we are planning to use does not seem to be like that, so that might simplify things. Then just adding the code change that you propose is sufficient.

Reply:
in real omop you are right it is more complicated; but in this harmonized limited omop we can make sense assumptions; in the case you are talking about we actually would get sex from teh PERSON table to be OMOP standard which would be a concept ID. If you look at the self-reported sex observation record you will see the `value_as_concept_id` is also there which would require a secondary lookup in CONCEPT table, but they filled in the string value of the concept for us.

Reply:
got it, thanks!